### ISSUE TITLE
[DOC] Firebase Authentication operations  - Multilayer login (WIP)

### ISSUE THREAD

**Issue by spoxies** [dpa99c/cordova-plugin-firebasex#640]
# Documentation suggestion

## What
Extending the documentation by showing how to use `cordova-firebasex-plugin` in conjunction with the `Firebase Web JS SDK`. 

## Why
The `cordova-firebasex-plugin` does things that the `Firebase Web JS SDK` cannot expose (nicely) within a hybrid environment (like auth/notifications/crashalytics). On the other hand the plugin isn't a full substitute of all (essential) features that the `Web JS SDK` exposes.

Since Firebase is already committed to [cordova compatibility](https://firebase.googleblog.com/2013/04/firebase-adds-phonegap-cordova-support.html) I would (advise a cordova developer to) use the `Web JS SDK` for everything where it does not needs to interact with the Native layer. Also keeping this `plugin` up with things that the `Firebase Web JS SDK` can support seems an everlasting/draining chase for the contributor(s). 

Since version 8.0.0. the auth has been [reworked](https://github.com/dpa99c/cordova-plugin-firebasex/commit/78bff37626fecd92ddc5de189de94e9864d87b6c). Meaning it is not possible/advised anymore to pass credentials to the `Web JS SDK` and login. There seem to be many issues/struggles (https://github.com/dpa99c/cordova-plugin-firebasex/issues/176, ) derived from this change and PR's addressing this are already posted (https://github.com/dpa99c/cordova-plugin-firebasex/pull/633). Documenting how to make use of these PR's does make sense.

## Why (not)/ Why warn
'Unfortunately' there is overlap like Firestore [functions](https://github.com/dpa99c/cordova-plugin-firebasex#firestore) that are (also) added in `cordova-plugin-firebasex`. Having the `Firebase Web JS SDK` (optionally put there by the app developer) as well, means that in this example there are multiple approaches to retrieve data. That might lead to mixing (native) variables and `Web JS SDK` variables. 

Reason being there are some essentials (and nice-to-haves) missing in `cordova-plugin-firebasex` and possibly for that same reason https://github.com/dpa99c/cordova-plugin-firebasex/pull/621 and https://github.com/dpa99c/cordova-plugin-firebasex/pull/633 where created. 

## Other
The above is based upon my current and limited understanding of this project and should by not be taken as truth.

### Sources
https://github.com/dpa99c/cordova-plugin-firebasex/pull/633
https://github.com/dpa99c/cordova-plugin-firebasex/commit/78bff37626fecd92ddc5de189de94e9864d87b6c

### Passing authentication to the Firebase JavaScript SDK
**Experimental:** Passing authentication to the [Firebase JavaScript SDK](https://firebase.google.com/docs/reference/js) is possible for some [Firebase Authentication operations](https://github.com/dpa99c/cordova-plugin-firebasex#authentication), but highly experimental. Please note that mixing variables from the native (plugin) and the JavaScript SDK should be avoided as it to will lead to unexpected behaviour. 

#### authenticateUserWithApple (Android/iOS)
Passing authentication unsupported (See https://github.com/dpa99c/cordova-plugin-firebasex/issues/551)

#### authenticateUserWithGoogle (Android/iOS)
```javascript
var clientId = '[secret].apps.googleusercontent.com';
// Trigger plugin native authentication
FirebasePlugin.authenticateUserWithGoogle(clientId, function (credential) {
    //Pass to Firebase JS SDK
    var { idToken } = credential; // Extract idToken
    firebase.auth.GoogleAuthProvider.credential(idToken);
}, function () {
    //err
});
```

#### verifyPhoneNumber (Android/iOS)  - w/o `instantVerification: true` 
NOTE: Passing authentication is unsupported if `instantVerification: true` (Android) is used (See https://github.com/dpa99c/cordova-plugin-firebasex/issues/176)

```javascript
var phonenumber;
var timeout;
FirebasePlugin.verifyPhoneNumber(function (credentials) {
    // Success
    if (credentials.instantVerification) {
        return alert('instantVerification not supported');
    }
    promptUserToInputCode() // you need to implement this
        .then(function (userEnteredSMSCode) {
            credential.code = userEnteredSMSCode;
            //Pass to Firebase JS SDK
            var webCredential = firebase.auth.PhoneAuthProvider.credential(credential, code);
            firebase.auth.signInWithCredential(webCredential);
        });
},
    function () {
        //err
    },
    phonenumber,
    timeout
);


```

**Comment by helpmetheroadassistanceapp** [comment]
You managed to find a solution? from my side, I could only use the web code on Android and the plugin code on iOS.

**Comment by spoxies** [comment]
@helpmetheroadassistanceapp Yes I just stared with finalising the Android app so I came across my own posts (and I over looked you question). Assuming this is a wide spread issue, I'm gonna reply months later for those who come across.

The actual solution/strategy turned out to be pretty simple. And when I find time I'll write a proposal on how to document this.

In short both the plugin and the web layer expose the `signInUserWithCustomToken` /  `signInWithCustomToken`. It is easy to generate such a token on firebase functions with an already logged in user. 

The resulting token you pass to the either one (web or native) you have not logged in to yet. So with an other token you can login as the same user no matter what initial authentication method you had chosen. 

Some semi pseudo code :

So what you do is expose an end point in firebase functions e.g:

**Firebase functions**
```javascript
// index.js
// HTTP :: /my/auth
app.get('/my/auth', (req, res) => {
    res.set('Content-Type', 'application/json');
    return api.myAuth_get(req, res);
});

app.options('/my/auth', (req, res) => {
    res.set('Content-Type', 'application/json');
    return api.myAuth_options(req, res);
});

```

```javascript
// == api.myAuth_options ===================================
exports.http_options = (req, res) => {
    // Set CORS headers
    res.set({
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'DELETE,GET,PATCH,POST,PUT',
        'Access-Control-Allow-Headers': 'Content-Type,Access-Control-Allow-Headers,Authorization,X-Requested-With',
    });

    res.status(200).send();

    return;
};
```



```javascript
// == api.myAuth_get ===================================
const admin = require('firebase-admin');

exports.http_get = (req, res) => {
    // Set CORS headers
    res.set({
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'DELETE,GET,PATCH,POST,PUT',
            'Access-Control-Allow-Headers': 'Content-Type,Access-Control-Allow-Headers,Authorization,X-Requested-With',
    });

    if ((!req.headers.authorization || !req.headers.authorization.startsWith('Bearer '))) {
        res.status(403).send('Unauthorized');
        return;
    } else {
        // Read the ID Token from the Authorization header.
        idToken = req.headers.authorization.split('Bearer ')[1];
    }

    return admin.auth().verifyIdToken(idToken)
        .then((decodedIdToken) => {
            return admin.auth().createCustomToken(decodedIdToken.uid);
        }).then((customToken) => {
            if (customToken) {
                res.status(200).send({'customToken': customToken});
                return;
            }
            res.status(404).send({
                'success': false,
            });

        }).catch((error) => {
            console.log('Error while creating a Firebase ID token:', error);
            res.status(400).send({ error: 'Invalid request' });
        });
};

```
So assuming you have logged in on the weblayer first you call the above endpoint (but you can do the other way around):

**APP:**
```javascript

firebase.auth().signInWithCredential(webCredential)
   .then((authUser) => {

      authUser.getIdToken().then(function (idToken) {
         return app.request.promise({
            url: [your - exposed - admin - endpoint]/my/auth,
            method: 'GET',
            dataType: 'json',
            headers: { 'Authorization': `Bearer ${idToken}` },
         }).then(function (customTokenData) {
            // Signs in the native layer
            window.FirebasePlugin.signInUserWithCustomToken(
               customTokenData.customToken,
               console.log,
               console.warn
            );
         });
      });
   });
```

**Comment by spoxies** [comment]
@dpa99c Would the above (but reworked a bit), be something that you like to have added to the docs 'Achieve multilayer login' ? If so I'll write a bit and propose a PR.

**Comment by dpa99c** [comment]
@spoxies Yes, any additional documentation to assist other plugin users for common use cases like this would be most welcome and appreciated.
